tokenizer_ts/src/bytePairEncode.ts

```diff
@@ -7,31 +7,48 @@ const enum Constant {
   BytesPerLevel = 6,
 }
 
-const binaryMapKey = (k: Uint8Array): number => {
-  const lower = k[0] | (k[1] << 8) | (k[2] << 16);
-  const upper = 0xFFFFFF * (k[3] | (k[4] << 8) | (k[5] << 16));
-  return lower + upper;
+const binaryMapKey = (k: Uint8Array, start: number, end: number): number => {
+  const length = end - start;
+
+  // 'lower' and 'upper' are both 24-bit integers, like
+  //    0xFF FF FF
+  //      ^3 ^2 ^1
+  // If we say have a length of 2, we should disregard the last "3" byte, so we
+  // create a mask like
+  //    0x00 FF FF  (started at 0xFF FF FF and shifted over by 8 bits)
+  //      ^3 ^2 ^1
+  // so that we discard the data outside our range
+  const lowerMask = 0xFFFFFF >>> Math.max(0, (3 - length) * 8);
+  const lower = (k[start + 0] | (k[start + 1] << 8) | (k[start + 2] << 16)) & lowerMask;
+  
+  if (length <= 3) {
+    return lower;
```

Comment:
There was a bug where the upperMask was mistakenly `>0` if length `<=3`

Reply:
TIL that if you try to right shift more than 31 bits it seems to wrap, `0xFFFFFF >>> 32 = 0xFFFFFF`.

I think you can fix this without a conditional by having

```
const upperMask = 0xFFFFFF >>> Math.min(31, Math.max(0, (6 - length) * 8)));
```

which might be faster, I'm not sure.

Reply:
I was thinking that adding the return statement was faster, since it skips the extra arithmetic for high bits?

Reply:
It could be, at the expense of a branch. Bitwise stuff is pretty speedy, branches can be pretty expensive. I'm not sure whether the difference would be enough to notice using the performance notebook (at least not without cranking up the minimum time to something very long)

Reply:
I changed to have no branch because it actually made it much slower 
![image](https://github.com/andreamah/Tokenizer/assets/31675041/bc6d015c-bae8-4e01-89a0-5ee90bc1ce17)

🙈

Reply:
Oh dang I didn't expect the difference to be that radical. Cool to see in practice though 😛

Reply:
Yeah, it's so strange haha